core/src/main/java/org/infinispan/interceptors/locking/AbstractTxLockingInterceptor.java

```diff
@@ -109,16 +101,60 @@ protected final Object invokeNextAndCommitIf1Pc(TxInvocationContext ctx, Prepare
     * locks to be released. The backup lock will be released either by a commit/rollback/unlock command or by
     * the originator leaving the cluster (if recovery is disabled).
     */
-   protected final void lockAndRegisterBackupLock(TxInvocationContext ctx, Object key, long lockTimeout, boolean skipLocking) throws InterruptedException {
-      //with DeltaCompositeKey, the locks should be acquired in the owner of the delta aware key.
+   protected final void lockOrRegisterBackupLock(TxInvocationContext<?> ctx, Object key, long lockTimeout, Action action)
+         throws InterruptedException {
       Object keyToCheck = key instanceof DeltaCompositeKey ?
             ((DeltaCompositeKey) key).getDeltaAwareValueKey() :
             key;
-      if (cdl.localNodeIsPrimaryOwner(keyToCheck)) {
-         lockKeyAndCheckOwnership(ctx, key, lockTimeout, skipLocking);
-      } else if (cdl.localNodeIsOwner(keyToCheck)) {
-         ctx.getCacheTransaction().addBackupLockForKey(key);
+      switch (LockUtil.getLockOwnership(keyToCheck, cdl)) {
+         case PRIMARY:
+            checkPendingAndLockKey(ctx, key, lockTimeout);
+            if (action != null) {
+               action.execute(ctx, key);
+            }
+            break;
+         case BACKUP:
+            ctx.getCacheTransaction().addBackupLockForKey(key);
+            break;
+         default:
+            break; //no-op
       }
+      ctx.addAffectedKey(key);
+   }
+
+   /**
+    * Same as {@link #lockOrRegisterBackupLock(TxInvocationContext, Object, long, Action)}
+    */
+   protected final void lockAllOrRegisterBackupLock(TxInvocationContext<?> ctx, Stream<Object> keys, long lockTimeout,
+                                                    Action action) throws InterruptedException {
+      final Log log = getLog();
+      final boolean trace = log.isTraceEnabled();
+
+      Collection<Object> keysToLock = keys.peek(ctx::addAffectedKey).filter(key -> {
+         Object keyToCheck = key instanceof DeltaCompositeKey ?
+               ((DeltaCompositeKey) key).getDeltaAwareValueKey() :
+               key;
+         switch (LockUtil.getLockOwnership(keyToCheck, cdl)) {
+            case PRIMARY:
+               if (trace) {
+                  log.tracef("Acquiring locks on %s.", key);
+               }
+               return true;
+            case BACKUP:
+               if (trace) {
+                  log.tracef("Acquiring backup locks on %s.", key);
+               }
+               ctx.getCacheTransaction().addBackupLockForKey(key);
+            default:
+               return false;
+         }
+      }).collect(Collectors.toList());
+
+      checkPendingAndLockAllKeys(ctx, keysToLock, lockTimeout);
+      if (action != null) {
+         keysToLock.forEach(key -> action.execute(ctx, key));
```

Comment:
then the caller has to filter the keys again. otherwise, I had to copy the filtering code everywhere before this method is invoked.